tests/pipelines/stable_diffusion/test_stable_diffusion.py

```diff
@@ -1423,3 +1424,56 @@ def test_stable_diffusion_euler(self):
         )
         max_diff = np.abs(expected_image - image).max()
         assert max_diff < 1e-3
+
+
+# (sayakpaul): This test suite was run in the DGX with two GPUs (1, 2).
+# The `test_components_put_in_right_devices` assertions depend on the device
+# on which the test is being run.
+@slow
+@require_torch_multi_gpu
+class StableDiffusionPipelineDeviceMapTests(unittest.TestCase):
+    def tearDown(self):
+        super().tearDown()
+        gc.collect()
+        torch.cuda.empty_cache()
+
+    def get_inputs(self, generator_device="cpu", seed=0):
+        generator = torch.Generator(device=generator_device).manual_seed(seed)
+        inputs = {
+            "prompt": "a photograph of an astronaut riding a horse",
+            "generator": generator,
+            "num_inference_steps": 50,
+            "guidance_scale": 7.5,
+            "output_type": "np",
+        }
+        return inputs
+
+    def test_forward_pass_balanced_device_map(self):
+        sd_pipe = StableDiffusionPipeline.from_pretrained(
+            "runwayml/stable-diffusion-v1-5", torch_dtype=torch.float16
+        ).to(torch_device)
+        sd_pipe.set_progress_bar_config(disable=True)
+        inputs = self.get_inputs()
+        no_device_map_image = sd_pipe(**inputs).images
+
+        del sd_pipe
+
+        sd_pipe_with_device_map = StableDiffusionPipeline.from_pretrained(
+            "runwayml/stable-diffusion-v1-5", device_map="balanced", torch_dtype=torch.float16
+        )
+        sd_pipe_with_device_map.set_progress_bar_config(disable=True)
+        inputs = self.get_inputs()
+        device_map_image = sd_pipe_with_device_map(**inputs).images
+
+        max_diff = np.abs(device_map_image - no_device_map_image).max()
+        assert max_diff < 1e-3
+
+    def test_components_put_in_right_devices(self):
+        sd_pipe_with_device_map = StableDiffusionPipeline.from_pretrained(
+            "runwayml/stable-diffusion-v1-5", device_map="balanced", torch_dtype=torch.float16
+        )
+
+        assert str(sd_pipe_with_device_map.unet.device) == "cuda:1"
+        assert str(sd_pipe_with_device_map.vae.device) == "cuda:1"
+        assert str(sd_pipe_with_device_map.safety_checker.device) == "cuda:0"
+        assert str(sd_pipe_with_device_map.text_encoder.device) == "cuda:0"
```

Comment:
since we are sorting the gpus by available memory, this assert can be flaky. Maybe just check that the device_map `generated` contrains at least 2 gpus ("cuda:1","cuda:2"). Also it can be useful to save the `device_map` generated by setting it as an attribute. In transformers, we have `model.hf_device_map` when a model is loaded with a device_map. You could do the same for a pipeline.

Reply:
Brilliant suggestion. Do the changes in 9ba3773b make sense?

Reply:
makes sense !